docs/design/namespaces.md

```diff
@@ -1,193 +1,219 @@
+A *LimitRange* defines min/max constraints on the amount of resources a single entity can consume in
+a *Namespace*.  See [Admission control: Limit Range](admission_control_limit_range.md)
 
-| Action | HTTP Verb | Path | Description |
-| ---- | ---- | ---- | ---- |
-| CREATE | POST | /api/{version}/{resourceType}/ | Create instance of {resourceType} in namespace *default* |
-| GET | GET | /api/{version}/{resourceType}/{name} | Get instance of {resourceType} in namespace *default* |
-| UPDATE | PUT | /api/{version}/{resourceType}/{name} | Update instance of {resourceType} in namespace *default* |
-| DELETE | DELETE | /api/{version}/{resourceType}/{name} | Delete instance of {resourceType} in namespace *default* |
+A *ResourceQuota* tracks aggregate usage of resources in the *Namespace* and allows cluster operators
+to define *Hard* resource usage limits that a *Namespace* may consume.
+See [Admission control: Resource Quota](admission_control_resource_quota.md)
 
-The k8s API server verifies the *Namespace* on resource creation matches the *{ns}* on the path.
+### Finalizers
 
-The k8s API server will enable efficient mechanisms to filter model resources based on the *Namespace*.  This may require
-the creation of an index on *Namespace* that could support query by namespace with optional label selectors.
+Upon creation of a *Namespace*, the creator may provide a list of immutable *Finalizer* objects.
 
-The k8s API server will associate a resource with a *Namespace* if not populated by the end-user based on the *Namespace* context
-of the incoming request.  If the *Namespace* of the resource being created, or updated does not match the *Namespace* on the request,
-then the k8s API server will reject the request.
+```
+// NamespaceSpec describes the attributes on a Namespace
+type NamespaceSpec struct {
+  // Finalizers is a list of named finalize actions that must complete prior to namespace deletion.
+  Finalizers []FinalizerName
+}
 
-TODO: Update to discuss k8s api server proxy patterns
+type FinalizerName string
 
-## k8s storage
+// These are internal finalizers to Kubernetes, must be qualified name unless defined here
+const (
+  FinalizerKubernetes FinalizerName = "kubernetes"
+)
 
-A namespace provides a unique identifier space and therefore must be in the storage path of a resource.
+// NamespaceStatus is information about the current status of a Namespace.
+type NamespaceStatus struct {
+  // Phase is the current lifecycle phase of the namespace.
+  Phase NamespacePhase `json:"phase,omitempty"`
+  // Finalizers is the list of named finalize actions that have completed
+  Finalizers []FinalizerName `json:"finalizers"`
+}
+```
 
-In etcd, we want to continue to still support efficient WATCH across namespaces.
+A *FinalizerName* is a qualified name.
 
-Resources that persist content in etcd will have storage paths as follows:
+A *Finalizer* denotes a named entity that must release it's claim to the *Namespace* resource
```

Comment:
This is the name of an object? Why not an ObjectReference?

Finalizers are expected to be objects that are external to the namespace?

Could you provide an example?